persistence/src/main/java/com/epam/esm/persistence/dao/impl/CertificateDAOImpl.java

```diff
@@ -0,0 +1,124 @@
+package com.epam.esm.persistence.dao.impl;
+
+import com.epam.esm.persistence.constants.CertificateQuerries;
+import com.epam.esm.persistence.dao.CertificateDAO;
+import com.epam.esm.persistence.mapper.CertificateMapper;
+import com.epam.esm.persistence.util.EntityFinder;
+import com.epam.esm.model.entity.Certificate;
+import com.epam.esm.persistence.exceptions.DAOSQLException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Component;
+import java.sql.*;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Component
+public class CertificateDAOImpl implements CertificateDAO {
+    static Logger logger = LoggerFactory.getLogger(CertificateDAOImpl.class);
+
+    private JdbcTemplate template;
+    private CertificateMapper certificateMapper;
+
+    @Autowired
+    public CertificateDAOImpl(JdbcTemplate template, CertificateMapper certificateMapper) {
+        this.template = template;
+        this.certificateMapper = certificateMapper;
+    }
+
+    public void setTemplate(JdbcTemplate template) {
+        this.template = template;
+    }
+
+    public JdbcTemplate getTemplate() {
+        return this.template;
+    }
+
+    @Override
+    public Certificate create(Certificate certificate) throws DAOSQLException {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+
+        template.update(connection -> {
+            PreparedStatement ps = connection
+                    .prepareStatement(CertificateQuerries.INSERT_QUERY.getValue(),
+                            Statement.RETURN_GENERATED_KEYS);
+            ps.setString(1, certificate.getName());
+            ps.setString(2, certificate.getDescription());
+            ps.setBigDecimal(3, certificate.getPrice());
+            ps.setInt(4, certificate.getDuration());
+            ps.setDate(5, Date.valueOf(certificate.getCreateDate()));
+            ps.setDate(6, Date.valueOf(certificate.getLastUpdateDate()));
+            return ps;
+        }, keyHolder);
+        if (keyHolder.getKey() == null) {
+            throw new DAOSQLException("empty keyholder");
+        }
+        certificate.setId(keyHolder.getKey().intValue());
+        return certificate;
+    }
+
+    @Override
+    public Certificate read(int id) throws DAOSQLException {
+        List<Certificate> result = template.query(CertificateQuerries.READ_QUERY.getValue()
+                .concat(CertificateQuerries.WHERE_ID.getValue()
+                        .replace("?", Integer.toString(id))),
+                certificateMapper);
+        if(result.isEmpty()) {
```

Comment:
you can return Optional to avoid this check

Reply:
is it necessary? It will cost me an addtitional variable and additional code line.

Reply:
could you please clarify what additional variable and line do you mean?
it should look something like this:

return template.query(CertificateQuerries.READ_QUERY.getValue().concat(CertificateQuerries.WHERE_ID.getValue()),
                certificateMapper, id).stream().findFirst();

Reply:
Oh, I thought you mean use Optional.isPresent instead of (result.isEmpty().
Changed to 
return template.query(CertificateQueries.READ_QUERY.getValue()
                        .concat(CertificateQueries.WHERE_ID.getValue()),
                certificateMapper, id).stream().findFirst().get();

Reply:
.findFirst() throws exception if result is empty. So i have to return to the previous version